Add Skeletal Support for API V2 Implementation

michaelestrin opened the pull request:
Add Skeletal Support for API V2 Implementation

Layered skeletal support for API V2 implementation and acceptance
testing.  Generic use-case-specific endpoint and batch endpoint
controller implementation.  Common cross-service ping, batch,
metrics, and test controller implementation (and related acceptance
tests). Common ping, metrics (with domain service), version, and
test use-case implementation. Validation support. Middleware 
support.

Added command line --debug flag and related middleware
implementation to log all requests and responses.

Acceptance tests require addition of REPO_ROOT environment
variable whose value is set to the absolute path to the root
of the edgex-go repository source; required to find default
configuration.toml files when starting services for acceptance
testing.

Dependent on https://github.com/edgexfoundry/edgex-go/pull/2322, https://github.com/edgexfoundry/edgex-go/pull/2324.

Fixes: https://github.com/edgexfoundry/edgex-go/issues/2277

Signed-off-by: Michael Estrin <m.estrin@dell.com>

michaelestrin commented:
This PR introduces Golang-based acceptance tests.  Each test starts an instance of the service it is testing in a separate goRoutine in effectively the same way the service is started when it's run from a command line.  It leverages the default configuration.toml files that are defined in the various sub-directories under /cmd.  Unfortunately, the way we execute `go test` in the builds means the directory containing the test becomes the working directory.  I needed a way to point to the repository's /cmd directory to be able to pass a reference into the service for the directory containing the configuration files.

I accomplished this by adding a `REPO_ROOT` environment variable whose value is set to the absolute path of the root of the local copy of the edgex-go repository.  This environment variable (and an appropriate value) will need to be added to the edgex-go builds that execute `go test`.  

Without it, the builds will fail because the tests fail since we don't have an absolute path we can use to point at the configuration.toml files:

![image](https://user-images.githubusercontent.com/464327/72520625-723d8f00-381f-11ea-839d-d610569a6634.png)

@lranjbar How difficult will adding this be and what's the process for doing so?

michaelestrin commented:
Moving to this issue yesterday's related exchange in EdgeXFoundry's #devops channel with @ernestojeda on the need to add `REPO_ROOT` environment variable to the build:

![image](https://user-images.githubusercontent.com/464327/72664978-79a09c00-39c9-11ea-83e7-bc0288226921.png)

https://edgexfoundry.slack.com/archives/CE46S51DX/p1579300400003600

lranjbar commented:
@michaelestrin After looking into it looks like this will need some testing to implement. I just created an issue in ci-management to track.

michaelestrin commented:
Related - https://github.com/edgexfoundry/ci-management/issues/547

lranjbar commented:
recheck

lranjbar commented:
@michaelestrin The current Jenkins builds have been updated to support this new environment variable.

michaelestrin commented:
> @michaelestrin The current Jenkins builds have been updated to support this new environment variable.

Awesome!  Thank you @lranjbar!

michaelestrin commented:
Dependent upon https://github.com/edgexfoundry/edgex-go/issues/2321

michaelestrin commented:
Significant changes have been since the presentation in the Core Working Group this past week (additional refactoring; validation added; version endpoint support; alignment with specification for version, metrics, ping DTOs; etc.).

While there is much related follow-on work to do (version, kind, action acceptance test variations; add configuration endpoint; add failure path acceptance tests for each common endpoint; extract common code in acceptance tests into test harness; routable delegate unit test; Router unit test; handle unit test; debugging middleware unit test; use-case unit tests, etc.), this implementation demonstrates the basic architectural implementation for APIv2.

Added reviewers.  Will update PR from draft status when reviews are complete and community consensus is reached to merge the code.

cloudxxx8 commented on internal/pkg/v2/ui/http/api/common/batch/batch.go:
```diff
@@ -0,0 +1,55 @@
+)
+
+const (
+	Endpoint = "/api/batch"
```
In swagger document, batch is still under /api/v2, and I suggest we still give batch api a version in the path for the following reason:

1. It might be a problem if users can mixed different versions of request, especially we will remove v1 API in the future.
2. After getting more user feedback, we might still want to modify the design of batch request in the future, especially about the error handling.  Without the version in the path, we might be easy to encounter backward compatibility issue, although the current design looks good so far.

michaelestrin replied:
> In swagger document, batch is still under /api/v2

No, it isn't.  See https://github.com/edgexfoundry/edgex-go/blob/7ca3df59c01c7f56d3c83a840c5995cdb7718c20/api/openapi/v2/core-command.yaml#L302.  See also https://github.com/edgexfoundry/edgex-go/pull/2309#discussion_r368939010.

> I suggest we still give batch API a version in the path

The batch endpoint is our interface for alternate transport implementations -- think of asynchronous messaging.  The intent is for it to be able to take various different major-release versions of request DTOs going forward; i.e. a v2 use-case request and a v3 use-case request within the same transport request.

michaelestrin replied:
I disagree with versioning the batch endpoint.  Explanation follows.

> In swagger document, batch is still under /api/v2

No, it isn't.  See https://github.com/edgexfoundry/edgex-go/blob/7ca3df59c01c7f56d3c83a840c5995cdb7718c20/api/openapi/v2/core-command.yaml#L302.  See also https://github.com/edgexfoundry/edgex-go/pull/2309#discussion_r368939010.

> I suggest we still give batch API a version in the path

The batch endpoint is our interface for alternate transport implementations -- think of asynchronous messaging.  The intent is for it to be able to take various different major-release versions of request DTOs going forward; i.e. a v2 use-case request and a v3 use-case request within the same transport request.  As such, it makes no sense to version the URL.

Currently, the envelopes associated with the batch endpoint are not versioned.  This is intentional.  Consider that they currently include an implicit version field and number (that doesn't correspond to the API's version number).  If it becomes necessary to modify the envelope structure in the future, the intent is to add an optional explicit version field (and codify the existing implied version number) to support the changes.

cloudxxx8 replied:
Got it.  Thanks for the elaboration.

lenny-intel commented on internal/core/command/init.go:
```diff
@@ -75,3 +85,21 @@ func (b *Bootstrap) BootstrapHandler(_ context.Context, _ *sync.WaitGroup, _ sta
+		dic,
+		b.muxRouter,
+		handlers,
+		common.V2Routes(
```
Not clear how the Core Command specific routes are loaded. Might be helpful to explain in a comment in each of these loadV2Routes funcs.

michaelestrin replied:
This will become self-evident once the follow-on service-specific endpoint example is delivered.  For the record, service-specific controllers are included in a slice (currently empty) passed to `common.V2Routes()` whose output is subsequently passed to `routing.Initialize()`.

lenny-intel commented on internal/core/command/main.go:
```diff
@@ -43,9 +43,10 @@ import (
+	var debugMode, useRegistry bool
 	var configDir, profileDir string
 
+	flag.BoolVar(&debugMode, "debug", false, "Turns on request/response debug logging.")
```
Should I add this new flag to the new Common Commandline abstraction?

michaelestrin replied:
Not unless you want your PR to take a dependency on this one.  

Ultimately, I think it should be.  However, I'm not sure whether the other consumers of go-mod-bootstrap have any corresponding use or implementation associated with it at this point.  Which is my way of saying, the right time to make it a generic flag is when everyone using go-mod-bootstrap has a corresponding usage/implementation for it.

lenny-intel commented on internal/pkg/test/request.go:
```diff
@@ -0,0 +1,42 @@
+// Marshal is a factory method that returns json-marshalled content.
+func Marshal(t *testing.T, content interface{}) []byte {
+	body, e := json.Marshal(content)
+	assert.Nil(t, e)
```
Seems like this should be fatal, but assert isn't fatal if it fails. I think you should you check if it failed and then call t.fatal() if it did?
```
if !assert.Nil(t, e) {
  t.fatal()
}

michaelestrin replied:
Point taken.  These were intended as gates to help troubleshoot an issue that should never really occur.  I'll adjust the code.

lenny-intel commented on internal/pkg/test/request.go:
```diff
@@ -0,0 +1,42 @@
+	w := httptest.NewRecorder()
+	r, e := http.NewRequest(method, url, bytes.NewReader(body))
+	router.ServeHTTP(w, r)
+	assert.Nil(t, e)
```
Same comment about this should be fatal if it fails

michaelestrin replied:
Point taken.  These were intended as gates to help troubleshoot an issue that should never really occur.  I'll adjust the code.

lenny-intel commented on internal/pkg/v2/application/delegate/executable.go:
```diff
@@ -0,0 +1,52 @@
+	}
+}
+
+// apply creates a closure chain that wraps the provided ui.Executable with the provided handlers' behavior.
```
should `ui.Executable` be `application.Executable`

lenny-intel commented on internal/pkg/v2/application/delegate/executable.go:
```diff
@@ -0,0 +1,52 @@
+	return execute
+}
+
+// Execute defines the signature of an arbitrary executable unit of behavior; fulfills Executable contract.
```
Is `defines` the correct word here? It is actually doing the execute, correct?

michaelestrin replied:
Comment was cut/paste from contract.  Will update.

lenny-intel commented on internal/pkg/v2/application/dto/v2dot0/error/response.go:
```diff
@@ -0,0 +1,29 @@
+}
+
+// NewResponse is a factory function that returns an initialized Response struct.
+func NewResponse(baseResponse *dtoBase.Response) *Response {
```
How is this adding anything error wise to the base response?

michaelestrin replied:
Strictly speaking, it isn't.  However, in the OpenAPI specification the ErrorResponse is a separate entity from the BaseResponse.  This is what's being modeled in the code.  It provides for possible future changes to ErrorResponse without incurring the cost of refactoring the code base to add it in a later revision.

lenny-intel commented on internal/pkg/v2/application/usecases/common/metrics/metrics.go:
```diff
@@ -0,0 +1,58 @@
+}
+
+// Request returns an empty request associated with this use case.
+func (_ *UseCase) Request() interface{} {
```
Why is this needed?

lenny-intel replied:
How is this even called?

michaelestrin replied:
It's part of the Routable contract.

It enables the generic handling of inbound requests in batch.go and usecase.go.

It provides a concrete structure typed as an empty interface that the generic handlers use to unmarshal request DTOs.

To see where it's called, look for usages of the Request method in the Routable interface (routable.go).  Currently, it's called from four separate places:

![image](https://user-images.githubusercontent.com/464327/73545361-d441eb00-4400-11ea-8535-7f04d4ea3b47.png)

lenny-intel replied:
I have never seen this syntax `(_ *UseCase)` using the underscore. Get it now, THX.

michaelestrin replied:
I misunderstood your original question.  :)

lenny-intel commented on internal/pkg/v2/application/validator/common/common.go:
```diff
@@ -0,0 +1,55 @@
+func Validate(
+	r interface{},
+	assertionOK bool,
+	getRequestID getRequestID,
```
Why must this be a function rather than just pass in the requestID

michaelestrin replied:
Precedence.

Validate function is called immediately after a type assertion intended to convert a request (of type interface{}) to a use-case-specific request DTO.  It's generic code (and therefore has no knowledge of the concrete DTO the request is being asserted into.

Further, the first action of Validate is to test whether the type assertion was successful.

The follow-on validation -- which is generic in nature -- requires the requestID from the request being validated.  There's no other way to pass that into Validate without duplicating the common code to verify the type assertion result.  For example, it isn't safe to do this (from validator/version/version.go)...

```
	request, ok := r.(*dto.Request)
	response, status = common.Validate(
		r,
		ok,
                request.RequestID,
		behavior,
		application.NewBehaviorSlice(application.Version2, application.KindVersion, application.ActionCommand),
	)
```

...because we don't yet know the value of request.  It could be nil -- which would cause a panic.  It's only after we ensure the type assertion succeeded that we can leverage the asserted types instance to access its properties.

lenny-intel commented on internal/pkg/v2/application/usecases/common/version/version.go:
```diff
@@ -0,0 +1,43 @@
+// Execute encapsulates the behavior for this use case.
+func (uc *UseCase) Execute(r interface{}) (interface{}, application.Status) {
+	request, _ := r.(*dto.Request)
+	return dto.NewResponse(dtoBase.NewResponseForSuccess(request.RequestID), application.VersionLatest),
```
The version endpoint is suppose to return the service's release version, i.e. 1.1.0 or when this released 2.0.0, not the API version.

michaelestrin replied:
The specification for VersionResponse.version states "[the latest version supported by this service](https://github.com/edgexfoundry/edgex-go/blob/d8a62da57cabe0c8b37a1721a5b06be65ae47022/api/openapi/v2/core-command.yaml#L266)."   

This is ambiguous -- you think service's release version and I think latest version of the API supported by the service.

There is value in having both.  Do you think I should extend the OpenAPI VersionResponse definition to include both?

lenny-intel replied:
I think this needs to be discussing in Core WG. The App SDK uses the V1 version to validate that the core service release version is compatible with the SDK version. i.e., the major values match.

michaelestrin replied:
@tsconn23 note request to discuss in working group above.

To clarify, I'm suggesting we return -- as separate properties -- both the service's version number (which for Geneva will be 1.x) and the latest major.minor version of the supported API (which for Geneva will be 2.0) in the APIv2 version endpoint.  No change would be made to the data returned by the existing V1 endpoint.

Yes, this is a breaking change.  But so is the entirety of APIv2.

lenny-intel replied:
Works for me.

michaelestrin replied:
(unresolved to keep on my radar for implementation.)

michaelestrin replied:
Discussed in today's Core WG call.

Version endpoint is not versioned in URL.

Will update v2 implementation to match existing v1 implementation.  Will remove v2's version endpoint to remove ../v2/.. from URI path.  Will take v2 version endpoint out of common routing wire-up for Geneva.

michaelestrin replied:
On further investigation, I'm not sure the above-mentioned approach will work well.  I'm going to leave the /api/v2/version endpoint in place.  I intend to defer decision on when/how to implement /api/version for after the Geneva release.

lenny-intel commented on internal/pkg/v2/ui/common/routable.go:
```diff
@@ -0,0 +1,27 @@
+ * the License.
+ *******************************************************************************/
+
+package common
```
Why `ui` rather than `api` in the package path?

michaelestrin replied:
It's a generic user-interface-level concern.

api is under http -- which is transport-specific.

It's placed outside of http in a common package because it is intended to be reused by future alternative transport implementations.

lenny-intel replied:
When i see `ui`, I have the expectation that there will be visual user interface code.... I think this will be confusing. Would `usecase` be more descriptive?

michaelestrin replied:
No, it would not.  Use-case is already in use; it's an application-level concept.

FWIW, I'm using the layers common to a domain-driven design application that uses traditional Layers Architecture [Buschmann, Frank, et al. 1996.  _Pattern-Oriented Software Architecture, Volume 1: A System of Patterns_. New York: Wiley.]. 

The user interface layer is simply the top-most layer that interacts with the external world; i.e. the user.  It's an extremely applicable label for the responsibilities assigned to this layer in this implementation.

It's unfortunate that user interface equates conceptually to a screen with windows and dialogs.  The term is actually quite generic:

![image](https://user-images.githubusercontent.com/464327/73568474-146b9280-442e-11ea-9041-8e529fc3925e.png)

lenny-intel replied:
Ok, you have sold me... :-)

lenny-intel commented on internal/pkg/v2/ui/common/routable/delegate.go:
```diff
@@ -0,0 +1,45 @@
+	}
+}
+
+// Execute defines the signature of an arbitrary executable unit of behavior.
```
Why `defines the signature`? When read this I expect a type rather than a func.

michaelestrin replied:
Comment was cut/paste from contract. Will update.

michaelestrin commented:
@lenny-intel I addressed all but one of your review comments.  Please unresolve any of them that require additional dialogue or clarification.  Thank you (!!!) for taking the time yesterday to walk through portions of the design and code with me and to review this PR.

michaelestrin commented:
Rebased.  Added change set (https://github.com/edgexfoundry/edgex-go/pull/2285/commits/864fa71329cfc98f57020ec12ee415cb83c9ab41) to address @lenny-intel's review comments and improve acceptance testing implementation (and unhappy path coverage).

michaelestrin commented:
> GoKit

Beyond the name, I know nothing of GoKit.  Not sure how it stacks up to our requirements.  This is something to be raised at the next Core working group meeting.  I'd ask that we refrain from discussing alternatives within the context of this PR.

michaelestrin commented:
> My biggest concern with this is the high level of complexity and the ability to be maintained by others than @michaelestrin .

I share this concern.  The last thing I want to do is deliver an architecture and implementation that isn't well understood by at least a few other people in the community.  That's the path to having the architecture and implementation quickly compromised.

My intent is to provide skeleton code, example use-case-specific implementation that leverages persistence (i.e. metadata addressable), and a follow-on ADR.  Based on your and @AnthonyMBonafide's suggestion, it would make sense to provide summary guidance on how to add an endpoint to the system.  Once these deliverables are complete, the community should have what it needs to begin to contribute APIv2 implementation.

It's important to note that you don't need to understand all of the complexity of the framework to add new functionality.  You just need to know how new functionality hangs off the framework.

michaelestrin commented on internal/pkg/v2/ui/http/handle/usecase.go:
```diff
@@ -0,0 +1,109 @@
+	}
+
+	request := routable.Request()
+	if request == nil {
```
I think this is answered by https://github.com/edgexfoundry/edgex-go/pull/2285#discussion_r373496056 ...?

lenny-intel replied:
routable.Request() returns a new empty Request correct? So why would it ever be nil? And why would you want to execute on a nil Request.

michaelestrin replied:
Yes, routable.Request() returns a new empty Request (that's a use-case specific concrete structure or nil).  

A nil request DTO simply specifies that the use-case doesn't expect or require a request DTO.  In this edge case, you'd still want to execute the invoked use-case but wouldn't want to attempt to unmarshal into a DTO because a DTO isn't expected or wanted.

Resolve if this addresses your question; ask another question if it does not.

lenny-intel replied:
Ok, that makes sense. All the example so far returned empty Request DTOs, so was expectaing that to be the standard, i.e. never a nil.

Please consider adding a comment with this `why` explanation

lenny-intel commented on internal/pkg/v2/ui/http/batchdto/common.go:
```diff
@@ -0,0 +1,57 @@
+ * the License.
+ *******************************************************************************/
+
+// batchdto defines the DTO envelope that wraps a use-case DTO for batch requests.  Unlike use-case DTOs (which are
```
Only this one left... ;-)

michaelestrin replied:
You're absolutely right.  

Moved from v2/ui/http/batchdto to v2/ui/common/batchdto.

Thank you for helping me improve my deliverable!

lenny-intel replied:
Shows that I actual understood it. There is hope... ;-)

michaelestrin commented:
Rebased

AnthonyMBonafide commented on internal/pkg/v2/ui/http/handle/usecase.go:
```diff
@@ -0,0 +1,110 @@
+
+	// multiple requests?
+	requests := []*json.RawMessage{}
+	err = json.Unmarshal(body, &requests)
```
I wonder if there is a better way to do this. With this layout we are forcing the caller to put all information in the body of the request rather than leveraging the HTTP transport for information, commonly done in REST. I do not think GET requests that just require an ID should contain a HTTP request body, since the information is in the URI. I also think it is important to have an internal representation of the DTO which encapsulates the request information to be processed regardless of transport. My thought is can we add functionality here which allows us to pull information from places other than just the HTTP body so that our APIs are more REST friendly?

michaelestrin replied:
> I do not think GET requests that just require an ID should contain a HTTP request body

It isn't the intention to have HTTP GET or DELETE requests to require an HTTP request body.  At the moment, the existing implementation doesn't have an example of GET or DELETE HTTP method requests; this is something I will be adding with the follow-on service-specific use-case example issue.

Implementation of GET and DELETE requests will -- within the corresponding controllers -- translate transport-specific implementation (such as identity or other content embedded in the requested URL; a security token passed in an HTTP header; etc.) to and from individual use-case request and response DTOs.

Unresolve the conversation and ask additional questions if this is not an adequate explanation.

AnthonyMBonafide replied:
I phrased my comment very poorly. What I meant was that in situations that only require an ID like getting an entity, which is typically done with a `GET` with the ID in the URI, would have to be be changed to a POST with the ID in the body with this implementation. This is because `GET` does not have a body and therefore we cannot get any context for the request.

With that said, I still believe you addressed my core concern of adding functionality to support data coming in via different HTTP mechanisms such as URI, headers, and query parameters.

eno-intel commented on internal/core/data/init.go:
```diff
@@ -40,22 +45,27 @@ import (
 	configuration := dataContainer.ConfigurationFrom(dic.Get)
+	lc := container.LoggingClientFrom(dic.Get)
+
+	loadRestRoutes(b.muxRouter, dic)
```
in core command you changed this function to loadV1Routes(...)
for consistency, it would be better to change this for core data and metadata, etc as well

eno-intel commented on internal/core/metadata/init.go:
```diff
@@ -37,29 +42,33 @@ import (
+	lc := bootstrapContainer.LoggingClientFrom(dic.Get)
 	registryClient := bootstrapContainer.RegistryFrom(dic.Get)
 
+	loadRestRoutes(b.muxRouter, dic)
```
in core command you changed this function to loadV1Routes(...)
for consistency, it would be better to change this for core data and metadata, etc as well

michaelestrin commented:
Rebased and squashed.

jpwhitemn commented:
Per special security/core WG meeting to talk about this Bryon Nevis raised a good question that we need to incorporate into the requirements and then design/implementation:  

Memory exhaustion - very very large batch requests.  Should we not cap number of requests in a batch (what is a reasonable limit?) or limit request payload size (what is a reasonable limit?)?
On the call, it was suggested that we should build in caps and/or limits via configuration that can be changed (since the actual numbers may not be really know) and put that into the transport implementation layer (e.g. HTTP, MQTT, etc.) as it will or may differ per transport.

mkbhanda commented:
Thank you everyone for making progress on this item. Really appreciate Bryon seeding the discussion with solution ideas.

Jim, With respect to resource exhaustion, I raised at last week’s security WG call and Michael Estrin mentioned we may be able to use his pre and post filter hooks to provide user role specific quotas. definitely need to make this Configurable, including client specific variants because they may have radically different resource constraints(mobile versus server).
Further Jim you had suggested exploring OASIS to see if there were industry patterns to handle this. Need to dig more.
Along the way noticed https://www.f5.com/labs/articles/threat-intelligence/denial-of-service-vulnerabilities-discovered-in-http-2.
cheers
Malini


---
Sent from Workspace ONE Boxer<https://whatisworkspaceone.com/boxer>

On February 4, 2020 at 2:30:44 PM EST, Jim White <notifications@github.com> wrote:

Per special security/core WG meeting to talk about this Bryon Nevis raised a good question that we need to incorporate into the requirements and then design/implementation:

Memory exhaustion - very very large batch requests. Should we not cap number of requests in a batch (what is a reasonable limit?) or limit request payload size (what is a reasonable limit?)?
On the call, it was suggested that we should build in caps and/or limits via configuration that can be changed (since the actual numbers may not be really know) and put that into the transport implementation layer (e.g. HTTP, MQTT, etc.) as it will or may differ per transport.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fedgexfoundry%2Fedgex-go%2Fpull%2F2285%3Femail_source%3Dnotifications%26email_token%3DAAUTXVNWJ5SSVYKXLYVTUDTRBG66BA5CNFSM4KHR2LV2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEKY4OXQ%23issuecomment-582076254&data=02%7C01%7Cmbhandaru%40vmware.com%7C4d2bc7a65c5a4116d96608d7a9a8b8cc%7Cb39138ca3cee4b4aa4d6cd83d9dd62f0%7C0%7C0%7C637164414433174507&sdata=xmQfuWfoEN1aAqbF%2FnKW7PQoVndAOr3XAwd5ka4Xsxw%3D&reserved=0>, or unsubscribe<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAAUTXVKXGVGSUTPGYBAFNLDRBG66BANCNFSM4KHR2LVQ&data=02%7C01%7Cmbhandaru%40vmware.com%7C4d2bc7a65c5a4116d96608d7a9a8b8cc%7Cb39138ca3cee4b4aa4d6cd83d9dd62f0%7C0%7C0%7C637164414433184500&sdata=H%2BTthWfYGvZQv3VTEsk7TkYFCuz%2BxUo%2Fo2DtMm0LtUQ%3D&reserved=0>.

michaelestrin commented:
Rebased.

michaelestrin commented:
Rebased.

lenny-intel commented on internal/pkg/v2/README.md:
```diff
@@ -0,0 +1,275 @@
+
+Middleware is extensible and can be layered.  You can apply different middleware implementations to the same use-case; each middleware is implemented and executed in isolation from every other middleware.
+
+Middleware shares implementation with DTO validation; the underlying implementation to curry/wrap `application.Executable` behavior is common for both.
```
Not familiar with the term `curry/wrap`. What does this mean?

michaelestrin replied:
> In mathematics and computer science, currying is the technique of translating the evaluation of a function that takes multiple arguments into evaluating a sequence of functions, each with a single argument. For example, a function that takes two arguments, one from _X_ and one from _Y_, and produces outputs in _Z_, by currying is translated into a function that takes a single argument from _X_ and produces as outputs functions from _Y_ to _Z_.

https://en.wikipedia.org/wiki/Currying

lenny-intel replied:
I am very familiar with the technique, but never heard it referred to as Currying.

lenny-intel commented on internal/pkg/v2/README.md:
```diff
@@ -0,0 +1,275 @@
+1. An example of a service-specific endpoint needs to be implemented.  This will illustrate how controllers for HTTP GET and DELETE method-based endpoints will create use-case request DTOs from HTTP headers and/or URL content.  It will also illustrate the approach to persistence, domain models, and how persistence schema and DTO definition diverge.
+2. The APIv2 specifications are inconsistent.  Certain definitions related to the ping, metrics, configuration, and version endpoints are accurate.  The rest of the document is not.  I intend to update the documents to make them consistent after the service-specific endpoint example has been completed, reviewed, and merged.
+3. Creation of an ADR to document the design, architecture, and approach to the APIv2 implementation.  I intend to repurpose a good portion of this document to create the ADR after the service-specific endpoint example has been completed, reviewed, and merged.
+4. go-mod-core-contracts integration.
```
Mike and were discussing that the DTOs belong in go-mod-core-contracts. Thoughts?

michaelestrin replied:
Eventually, yes.  At least a copy of them as they constitute the contract.

michaelestrin commented on api/openapi/v2/core-command.yaml:
```diff
@@ -215,9 +215,9 @@ components:
           type: string
           description: "The name of the type applicable to the request instance contained in the 'content' property."
         version:
```
See https://github.com/edgexfoundry/edgex-go/pull/2285#discussion_r375950196 and https://github.com/edgexfoundry/edgex-go/pull/2285#discussion_r376013368.

michaelestrin commented on internal/pkg/v2/README.md:
```diff
@@ -0,0 +1,275 @@
+- ***DTO:*** data transfer object.  An object with no behavior used only for data transmission; e.g. the request and response objects that define the structure and content of the data required by and returned by a use-case.
+- ***Use-case Request:*** a request DTO instance that maps to a single invocation of a use-case.
+- ***Use-case Response:*** a response DTO instance returned from a single invocation of a use-case. 
+- ***Transport Request:***  a single HTTP request.  A single transport request can contain one or more use-case requests and an equivalent number of use-case responses.
```
I could.  I specified HTTP because it's what everyone is familiar with and provides a bridge from the v1 world they know to the new v2 world.

lenny-intel replied:
Ok, how about move it to the end as an example. Having it at the beginning is confusing ;-)

michaelestrin replied:
But the definition of a transport request is "a single HTTP request."  All that stuff afterwards talks about the relationship between transport requests and use-case requests.  Moving the definition to the end seems confusing to me.

lenny-intel replied:
So `Transport Request` is always going to be HTTP? Never a different transport? Then for clarity, how about changing it to `HTTP Transport Request`.

michaelestrin replied:
I really want to make you happy.  Really, I do.

The documentation is written to reflect the current state of things.  It is intended to bridge v2 to v1 concepts.  v1 was all about HTTP.  v2 carries that forward.  Transport Request is currently _only_ HTTP.  That's why I specify "a single HTTP request."  It bridges v2 and v1.  

Transport Request isn't solely an HTTP concept going forward.  One possible future implementation around asynchronous messaging is to provide a transport request (i.e. a single asynchronous message) that contains multiple use-case requests.  

Unresolve and comment if you want to continue this conversation here.

michaelestrin commented on internal/pkg/v2/README.md:
```diff
@@ -0,0 +1,275 @@
+- ***Use-case Request:*** a request DTO instance that maps to a single invocation of a use-case.
+- ***Use-case Response:*** a response DTO instance returned from a single invocation of a use-case. 
+- ***Transport Request:***  a single HTTP request.  A single transport request can contain one or more use-case requests and an equivalent number of use-case responses.
+- ***Use-case Endpoint:*** URL that accepts one or more use-case-specific requests and returns an equivalent number of use-case responses.  
```
I don't see an immediate equivalence between messaging topics and use-case endpoints.  You could implement a transport that way (topic per endpoint) but I don't know if doing so would be a recommended approach.

lenny-intel replied:
`Use-case Endpoint` is HTTP transport specific, so I would assume you would have something equivalent for a different transport, say Pu/Sub.

michaelestrin replied:
I'd expect alternate transports to have their own ubiquitous language terms and related concepts.

lenny-intel replied:
Yes, that is what I was try to confirm. :-)

michaelestrin commented on internal/pkg/v2/README.md:
```diff
@@ -0,0 +1,275 @@
+## v2 Package
+
+[TOC]
```
<sad face> 

I am using Typora.  It's a cool tool.  Thank you for pointing it out.

I had noticed that.  I don't have the cycles (or patience at this point) to manually create and update a TOC.  I know how to do it -- I'm just choosing not to at this point.  :)

michaelestrin replied:
FWIW - https://github.com/edgexfoundry/edgex-go/issues/2346

michaelestrin commented:
Rebased and addressed conflicts resulting from #2343 merge.

michaelestrin commented on internal/pkg/v2/README.md:
```diff
@@ -0,0 +1,275 @@
+
+Multiple use-case requests received in a single transport request must all be of same type (which corresponds to the request DTO expected by the underlying use-case).
+
+Multiple use-case requests received in a single transport request are processed concurrently.  Results are aggregated and returned as a single array of responses in the transport's response.  The relative position of a use-case response in the response's array has no correlation to the relative position of the corresponding use-case request in the request's array.
```
IMO, how requests and responses are correlated is an issue for the OpenAPI specification.

Unresolve and comment if you want to continue this conversation here.

lenny-intel replied:
Since you bring up the correlation issues, i.e. `response's array has no correlation to the relative position` it begs the question on how correlation is resolved.

michaelestrin commented on internal/pkg/v2/README.md:
```diff
@@ -0,0 +1,275 @@
+
+Multiple wrapped use-case requests received in a single transport request may be of mixed types.
+
+Multiple wrapped use-case requests received in a single transport request are processed sequentially.  Results are aggregated and returned as a single array of wrapped responses in the transport's response. The relative position of a wrapped use-case response in the response's array has no correlation to the relative position of the corresponding wrapped use-case request in the request's array.
```
> Add why

It was a somewhat arbitrary decision.  There's forward vision that would provide for the array of RequestEnvelope DTOs to be wrapped in an object that specified a processing strategy.  This would allow for much more complex saga-like interaction (i.e. validate that the array contains all of the items necessary to do some complex behavior; process all of the objects of type A concurrently; wait for type A processing to complete; process type B, type C, and type D objects sequentially and in that order; return a result).  You could also implement a strategy that provided for simple DTO-ignorant concurrent processing of all RequestEnvelope DTOs.

> Might want to add how the responses are match to the requests, i.e. requestId

IMO, how requests and responses are correlated is an issue for the OpenAPI specification.

Unresolve and comment if you want to continue this conversation here.

lenny-intel replied:
You gave me the background on the Why? But didn't address it in the document.

lenny-intel replied:
Since you bring up the correlation issues, i.e. response's array has no correlation to the relative position it begs the question on how correlation is resolved.

michaelestrin replied:
> You gave me the background on the Why? But didn't address it in the document.

I thought I did; it's in the document's "future" section.

![image](https://user-images.githubusercontent.com/464327/74238128-c27f0400-4c9a-11ea-9ecc-d27137c1f6ed.png)

> Since you bring up the correlation issues, i.e. response's array has no correlation to the relative position it begs the question on how correlation is resolved.

"IMO, how requests and responses are correlated is an issue for the OpenAPI specification."  It's a data/protocol-level concern and is addressed in the OpenAPI spec.

https://github.com/edgexfoundry/edgex-go/pull/2285#discussion_r376386873

michaelestrin commented:
Rebased.  

Added to documentation.  

Addressed test breakage resulting from https://github.com/edgexfoundry/edgex-go/issues/2340.

tonyespy commented on api/openapi/v2/core-command.yaml:
```diff
@@ -215,9 +215,9 @@ components:
           description: "The name of the type applicable to the request instance contained in the 'content' property."
         version:
-          description: "Proposed field for explicitly defining version of request DTO. This is for advertising compatibility between a publisher/subscriber or requester/receiver"
+          description: "Explicitly defines major-only version number of request DTO."
```
This implies DTOs also have minor (and possibly patch) numbers?

michaelestrin commented:
Rebased.  Addressed conficts resulting from merge of https://github.com/edgexfoundry/edgex-go/pull/2317

michaelestrin commented:
Rebased

michaelestrin commented:
Rebased and addressed conflicts created by https://github.com/edgexfoundry/edgex-go/pull/2351

michaelestrin commented:
Rebased and squashed.

michaelestrin commented:
Rebased.

tonyespy commented on internal/core/data/init.go:
```diff
@@ -116,3 +129,23 @@ func (b *Bootstrap) BootstrapHandler(_ context.Context, _ *sync.WaitGroup, _ sta
 }
+
+// loadV2Routes creates a new command-query router and handles the related mux.Router initialization for API V2 routes.
+func (b *Bootstrap) loadV2Routes(dic *di.Container, lc logger.LoggingClient) {
```
Other than the receiver, this method is identical across all the services in this PR. Perhaps this could be collapsed into a common function?

michaelestrin replied:
It's identical currently because there are no service-specific controllers being added.  

Longer-term vision is here -- https://github.com/michaelestrin/edgex-go/blob/issue-2277/internal/pkg/v2/README.md

> Single Configuration-Driven Executable
The design supports a single executable capable of supporting endpoints from one or more of today's existing services. This could be hardcoded -- as it is in the current Fuji release. It could also easily be configuration-driven -- a single executable could provide any or all existing functionality for all core, support, and system management services based on a configuration file. This would simplify artifact creation and provide a more flexible deployment story.

This would reduce code duplication across services significantly.

Regardless, there is still duplication -- like this -- in different places in the PR.  I'm evolving the design, identifying the duplication, and extracting (as I did within the last day or two with acceptance tests and creating test.Case) as it makes sense.  It doesn't make sense yet for me to extract this common code into one place.  It probably will after a few services have their controllers implemented and the full pattern becomes apparent.

tonyespy commented on internal/pkg/v2/application/dto/v2dot0/common/metrics/metrics.go:
```diff
@@ -0,0 +1,68 @@
+	dtoBase.Request `json:",inline"`
+}
+
+// Response defines the output/result for this use case.
```
gdoc for the fields would be helpful... I also was going to suggest as an alternative that a reference to the openapi definition for this DTO would be useful, but the question then is to which one, as this response DTO is defined in each service's openapi yaml file. Maybe we should investigate whether some of these common definitions can be extracted into their own yaml files and then reference by the main service yaml files?

tonyespy commented on internal/pkg/v2/infrastructure/test/acceptance.go:
```diff
@@ -0,0 +1,80 @@
+	method string,
+	versionVariations map[string][]*Case) {
+
+	for version := range versionVariations {
```
This function could use some additional comments...

michaelestrin replied:
I'm not sure what's not clear here so I'm not sure what additional comments are required.

tonyespy commented on internal/pkg/v2/ui/http/acceptance/common/metrics/v2dot0/cases.go:
```diff
@@ -0,0 +1,245 @@
+	"github.com/edgexfoundry/edgex-go/internal/pkg/v2/ui/common/batchdto"
+)
+
+// UseCaseTestCases returns a series of v2.0 test cases to test the ping use-case endpoint.
```
This func could use some additional comments...

michaelestrin replied:
I'm not sure what's not clear here so I'm not sure what additional comments are required.  Please provide some additional details @tonyespy.

tonyespy commented on internal/pkg/v2/ui/http/acceptance/common/ping/v2dot0/cases.go:
```diff
@@ -0,0 +1,294 @@
+
+// UseCaseTestCases returns a series of v2.0 test cases to test the ping use-case endpoint.
+func UseCaseTestCases(t *testing.T) []*test.Case {
+	return []*test.Case{
```
This func could use some additional comments...

michaelestrin replied:
I'm not sure what's not clear here so I'm not sure what additional comments are required.  Please provide some additional details @tonyespy.

tonyespy commented on internal/pkg/v2/README.md:
```diff
@@ -0,0 +1,296 @@
+
+### Overview
+
+This package encapsulates the APIv2 implementation.
```
This was super-helpful in understanding this PR. Thanks for including it, and the link to the related presentation!

tonyespy commented on internal/pkg/v2/README.md:
```diff
@@ -0,0 +1,296 @@
+
+APIv2 is being treated as greenfield development.  This reduces risk by minimizing overlap with APIv1 implementation.  It ensures APIv2 is able to be developed and evolve separately from APIv1.  It also provides a clear delineation between APIv1 and APIv2 and makes it easier (and less likely to introduce last-minute issues) to remove the APIv1 functionality for the future EdgeX 2.0 release.
+
+To this end, the community decided that APIv2 persistence will be separate from APIv1 persistence.  This supports independent schema development at the expense of data access across the two API versions.  That is, data persisted via APIv1 will not be available via APIv2 (and vice-versa).
```
I guess I missed this discussion. While I understand why it was made, I will point out that as all the current services use the V1 APIs, we can't really do any informal testing using v2 APIs to populate configuration, metadata, or runtime objects and then consume/validate them with v1 endpoints. Did we consider some sort of configuration flag to enable v2 endpoints and keep them using the same persistence back-end?

michaelestrin replied:
Decision was made in a recent working group call (I frankly can't remember which one at the moment) to keep v1 and v2 completely separate down to and including persistence.  No overlap whatsoever.  This enables us to evolve the v2 persistence schema independently of v1.    It also minimizes the risk of v2 implementation polluting v1 (and vice-versa) and the risk of impacting the v2 implementation when we eventually remove the v1 implementation.

tsconn23 replied:
FYI -- Decision was in Core WG 30-Jan.

tonyespy commented on internal/pkg/v2/README.md:
```diff
@@ -0,0 +1,296 @@
+
+The implementation spans 4 architectural layers:
+
+- ***User Interface Layer***  
```
I really dislike this term... I understand it's historical, but there's nothing "user interface" about this design.

michaelestrin replied:
https://github.com/edgexfoundry/edgex-go/pull/2285#discussion_r373254118

Unresolve if you want to revisit/continue this conversation.

michaelestrin commented:
> That said, this PR is huge, and frankly a bit overwhelming to review. 146 files (at last count), and thousands of lines of diff... I'm not advocating that this PR be broken up, but I'd really like us to consider how we can avoid these sorts of giant PRs every release cycle. I'd like to see us discuss this topic in our architects meeting.

FWIW, this is orthogonal infrastructure code that implements a completely new architecture.  Roughly 2/3 of the PR is test-related.  I'm in general agreement that smaller PRs are better, I'm not sure how I could have done this differently (and had the community be able to understand it) given the nature of the work.  Open to suggestions.

michaelestrin commented on api/openapi/v2/core-command.yaml:
```diff
@@ -215,9 +215,9 @@ components:
           description: "The name of the type applicable to the request instance contained in the 'content' property."
         version:
-          description: "Proposed field for explicitly defining version of request DTO. This is for advertising compatibility between a publisher/subscriber or requester/receiver"
+          description: "Explicitly defines major-only version number of request DTO."
```
An interesting observation; I didn't see it that way.

Intended to communicate that DTO's version field will contain only the DTO's major version.  All DTOs for a set of major version releases should be backward-compatible with prior minor releases.  As you may have seen in the code, DTOs are tracked in the code by major/minor release for acceptance testing to ensure backward-compatible implementation.

All that said, we're only specifying the major release version in the RequestEnvelope and ResponseEnvelope DTOs.

michaelestrin commented on internal/core/command/init.go:
```diff
@@ -23,12 +23,17 @@ import (
 	"github.com/edgexfoundry/edgex-go/internal/pkg/endpoint"
 	"github.com/edgexfoundry/edgex-go/internal/pkg/errorconcept"
+	"github.com/edgexfoundry/edgex-go/internal/pkg/v2/application/delegate"
+	"github.com/edgexfoundry/edgex-go/internal/pkg/v2/ui/common/middleware/debugging"
```
At some point, someone added `debug` to .gitignore; precludes naming the package "debug"

See https://github.com/edgexfoundry/edgex-go/blob/f7e219fb7c8b58b70b46411f0cd43839967fca39/.gitignore#L11

michaelestrin commented on internal/pkg/v2/ui/http/api/common/version/version.go:
```diff
@@ -0,0 +1,79 @@
+	version  = application.Version2
+	kind     = application.KindVersion
+	action   = application.ActionCommand
+	method   = api.ActionCommand
```
CRUD is reserved for resources; i.e. entities that you will create, read, update, and delete -- things that have a life-cycle and are stored in persistence.  Version isn't one of those; it's a command -- "give me the version."  There's no way to update the version.  No way to create a new one.  No way to delete it.  There's no persistence related to it.  There's only one version.

michaelestrin commented on internal/pkg/v2/infrastructure/test/response.go:
```diff
@@ -0,0 +1,29 @@
+	"github.com/stretchr/testify/assert"
+)
+
+// Unmarshal is a factory method that returns a structure.
```
It does.  I realize it isn't immediately apparent.  You pass in `t`and `content` and a reference to a `value` and we unmarshal the `content` into the `value`.  Because `value` is passed in as a reference, when the function returns it now contains (assuming there wasn't an error) the structure unmarshalled from `content`.

michaelestrin replied:
I will update the function's docblock comment.

michaelestrin commented on internal/core/command/init.go:
```diff
@@ -23,12 +23,17 @@ import (
+	"github.com/edgexfoundry/edgex-go/internal/pkg/v2/application/delegate"
+	"github.com/edgexfoundry/edgex-go/internal/pkg/v2/ui/common/middleware/debugging"
+	"github.com/edgexfoundry/edgex-go/internal/pkg/v2/ui/http/api/common"
+	"github.com/edgexfoundry/edgex-go/internal/pkg/v2/ui/http/routing"
```
I struggled a bit with this before renaming package to router (and renaming the internal router receiver).

michaelestrin commented on internal/pkg/v2/application/validator/metrics/metrics_test.go:
```diff
@@ -0,0 +1,83 @@
+		{
+			name:             "valid",
+			request:          dto.NewRequest(dtoBase.NewRequest(readingID)),
+			expectedResponse: dto.NewRequest(dtoBase.NewRequest(readingID)),
```
You really did review this code in detail!

This is a validator test.  It leverages `delegate.TestExecutable` to wrap the validation; if all validation passes, `delegate.TestExecutable.Execute()` returns the original request.  We test for responses in the other test cases because we expect an error response to be returned.  Here, we're not actually taking any action if everything validates -- we just check that what is returned (a request) is what we sent (a request).

michaelestrin commented on internal/pkg/v2/application/validator/ping/ping_test.go:
```diff
@@ -0,0 +1,83 @@
+		{
+			name:             "valid",
+			request:          dto.NewRequest(dtoBase.NewRequest(requestID)),
+			expectedResponse: dto.NewRequest(dtoBase.NewRequest(requestID)),
```
You really did review this code in detail!

This is a validator test. It leverages delegate.TestExecutable to wrap the validation; if all validation passes, delegate.TestExecutable.Execute() returns the original request. We test for responses in the other test cases because we expect an error response to be returned. Here, we're not actually taking any action if everything validates -- we just check that what is returned (a request) is what we sent (a request).

michaelestrin commented on internal/pkg/v2/application/validator/version/version_test.go:
```diff
@@ -0,0 +1,83 @@
+		{
+			name:             "valid",
+			request:          dto.NewRequest(dtoBase.NewRequest(requestID)),
+			expectedResponse: dto.NewRequest(dtoBase.NewRequest(requestID)),
```
You really did review this code in detail!

This is a validator test. It leverages delegate.TestExecutable to wrap the validation; if all validation passes, delegate.TestExecutable.Execute() returns the original request. We test for responses in the other test cases because we expect an error response to be returned. Here, we're not actually taking any action if everything validates -- we just check that what is returned (a request) is what we sent (a request).

michaelestrin commented:
Rebased and squashed.

michaelestrin commented on internal/pkg/v2/application/dto/v2dot0/ping/ping.go:
```diff
@@ -0,0 +1,60 @@
+	}
+}
+
+// NewEmptyRequest returns an uninitialized request structure for this use case.
```
It provides a concrete structure the generic handlers use to unmarshal a specific request's DTO.  I added corresponding comments to the code.

michaelestrin commented on internal/pkg/v2/application/status.go:
```diff
@@ -0,0 +1,29 @@
+// Status is the common type definition for the application-level transport-agnostic status code.
+type Status int
+
+const (
```
Yes, they are returned to client in the BaseResponse's statusCode property.  All use-case DTOs inherit/include BaseResponse.

Individual statusCode values have not been defined as part of the specification; they should (and will eventually) be.

I believe individual comments wouldn't help here -- the constant name should provide enough context to understand the status being communicated.

michaelestrin commented on internal/pkg/v2/infrastructure/test/timer.go:
```diff
@@ -0,0 +1,53 @@
+		return time.Duration(value)
+	}
+	elapsed := (*t.endTime).Sub(t.startTime) / time.Millisecond
+	return init(expectedInMS-deviationInMS) <= elapsed && elapsed < init(expectedInMS+deviationInMS)
```
It is complex but I'd posit succinct.  I added a clarifying comment.

michaelestrin commented on internal/pkg/v2/application/delegate/executable.go:
```diff
@@ -0,0 +1,55 @@
+	}
+}
+
+// apply creates a closure chain that wraps the provided application.Executable with the provided handlers' behavior.
```
It's arguably the most complex code in this PR.  Point taken; I added clarifying comments to the code.

michaelestrin commented:
Rebased.  Squashed.  Addressed most of @tonyespy's review comments.

michaelestrin commented:
Rebased.

lenny-intel commented on internal/pkg/v2/application/usecases/common/metrics/metrics.go:
```diff
@@ -0,0 +1,61 @@
+
+import (
+	dtoBase "github.com/edgexfoundry/edgex-go/internal/pkg/v2/application/dto/v2dot0/common/base"
+	dto "github.com/edgexfoundry/edgex-go/internal/pkg/v2/application/dto/v2dot0/common/metrics"
```
These generic alias make the code below harder to understand. Changing to `dtoMetrics` or `metricsDto` makes the following line from below more explicit that the DTO is the a Metrics DTO. 

```
return metricsDto.NewResponse
```
vs 

```
return dto.NewResponse
```
Same goes for `domain` vs `metricsDomain`

I don't have to think about what DTO or Domain it is when reading the code below if the aliases are more explicit.

michaelestrin replied:
I agree it makes the code more readable.

It also makes the patterns in the code less recognizable.  

I mentioned to you at some point in the last couple of weeks that this was an evolutionary architectural effort.  Which is to say, the design is evolving based on the patterns I detect in the code.  I'm dealing in generics.

On example of this is in the acceptance tests.  A common pattern became readily apparent from the various acceptance tests that were written.  This pattern was extracted within the past week or so into a single implementation (where before it was duplicated across many individual tests) and test.Case was created.

My goal here isn't to make the code more readable.  My goal is to structure the code to make the patterns more recognizable so common functionality can be extracted into its own implementation and the existing individual implementations of that functionality can be imploded.

Meanwhile, my ask is to pay attention to the file you're in (or that's aliased).  

Unresolve and comment if you want to continue this conversation.

michaelestrin commented:
Rebased.  Incorporated common changes required by follow-on https://github.com/edgexfoundry/edgex-go/issues/2312.  Moved validation into use-case New() factory function.
